src/ProgrammableDevice.cc

```diff
@@ -0,0 +1,103 @@
+#include "ProgrammableDevice.hh"
+
+#include "CommandController.hh"
+#include "MSXCliComm.hh"
+#include "MSXException.hh"
+#include "StringSetting.hh"
+#include "serialize.hh"
+
+#include "narrow.hh"
+
+namespace openmsx {
+
+ProgrammableDevice::ProgrammableDevice(const DeviceConfig& config)
+	: MSXDevice(config)
+	, portLayoutSetting(
+		getCommandController(), tmpStrCat(getName(), " port layout"),
+		"List of {port num} tuples are added to the pool of IO ports", "")
+	, resetCallback(getCommandController(),
+		getName() + " reset callback",
+		"Tcl proc to call when reset was detected",
+		"", Setting::Save::NO)
+	, readCallback(getCommandController(),
+		getName() + " read callback",
+		"Tcl proc to call when reading a port from the device",
+		"", Setting::Save::NO)
+	, writeCallback(getCommandController(),
+		getName() + " write callback",
+		"Tcl proc to call when writing to a port of the device",
+		"", Setting::Save::NO)
+{
+	portLayoutSetting.attach(*this);
+	readAllPorts(portLayoutSetting);
+	reset(getCurrentTime());
+}
+
+ProgrammableDevice::~ProgrammableDevice()
+{
+	portLayoutSetting.detach(*this);
+}
+
+void ProgrammableDevice::readAllPorts(const Setting& setting)
+{
+	Interpreter& interp = getCommandController().getInterpreter();
+	TclObject ports(((StringSetting&)setting).getString());
+
+	for (unsigned n = 0; n < ports.getListLength(interp); n++) {
+		TclObject item = ports.getListIndex(interp, n);
+		int port;
+		int num;
+		try {
+			port = item.getListIndex(interp, 0).getInt(interp);
+			num = item.getListIndex(interp, 1).getInt(interp);
+		} catch (...) {
+			continue;
+		}
+		inPorts.setPosN(port, num);
+		outPorts.setPosN(port, num);
+	}
+}
+
+void ProgrammableDevice::update(const Setting& setting) noexcept
+{
+	unregisterPorts();
+	inPorts = IterableBitSet<256>();
+	outPorts = IterableBitSet<256>();
+	readAllPorts(setting);
+	doRegisterPorts();
+}
+
+void ProgrammableDevice::writeIO(word port, byte value, EmuTime::param /*time*/)
+{
+	writeCallback.execute(port, value);
+}
+
+byte ProgrammableDevice::readIO(word port, EmuTime::param /* time */)
+{
+	byte result = 255;
+	try {
+		auto obj = readCallback.execute(port);
+		if (obj != TclObject()) {
+			auto tmp = obj.getInt(getCommandController().getInterpreter());
+			if ((tmp < 0) || (tmp > 255)) {
+			        throw MSXException("outside range 0..255");
+			}
+			result = narrow_cast<byte>(tmp);
+		}
+	} catch (MSXException& e) {
+		getCliComm().printWarning(
+			"Wrong result for callback function \"",
+			readCallback.getSetting().getFullName(),
+			"\": ", e.getMessage());
+	}
+	return result;
```

Comment:
Detail. I think the code becomes simpler with this structure:
```
try {
    ...
    return narrow_cast<byte>(tmp);
} catch (MSXException& e) {
    ...
    return 255;
}
```
So remove the local variable `result`, and have 2 instead of 1 `return` statements.